saliency/xrai.py

```diff
@@ -0,0 +1,555 @@
+        integer_segments: See XRAIParameters. Defaults to True.
+
+    Returns:
+        tuple: saliency heatmap and list of masks or an integer image with
+               area ranks depending on the parameter integer_segments.
+    """
+    output_attr = -np.inf * np.ones(shape=attr.shape, dtype=np.float)
+
+    n_masks = len(segs)
+    current_attr_sum = 0.0
+    current_area_perc = 0.0
+    current_mask = np.zeros(attr.shape, dtype=bool)
+
+    masks_trace = []
+    remaining_masks = {ind: mask for ind, mask in enumerate(segs)}
+
+    added_masks_cnt = 1
+    # While the mask area is less than area_th and remaining_masks is not empty
+    while current_area_perc <= area_perc_th:
+      best_gain = -np.inf
+      best_key = None
+      remove_key_queue = []
+      for mask_key in remaining_masks:
+        mask = remaining_masks[mask_key]
+        # If mask does not add more than min_pixel_diff to current mask, remove
+        mask_pixel_diff = _get_diff_cnt(mask, current_mask)
+        if mask_pixel_diff < min_pixel_diff:
+          remove_key_queue.append(mask_key)
+          if verbose > 2:
+            logging.info("Skipping mask with pixel difference: {:.3g},".format(
+                mask_pixel_diff))
+          continue
+        gain = gain_fun(mask, attr, mask2=current_mask)
+        if gain > best_gain:
+          best_gain = gain
+          best_key = mask_key
+      for key in remove_key_queue:
+        del remaining_masks[key]
+      if len(remaining_masks) == 0:
+        break
+      added_mask = remaining_masks[best_key]
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+      masks_trace.append((mask_diff, best_gain))
+
+      current_mask = np.logical_or(current_mask, added_mask)
+      current_attr_sum = np.sum(attr[current_mask])
+      current_area_perc = np.mean(current_mask)
+      output_attr[mask_diff] = best_gain
+      del remaining_masks[best_key]  # delete used key
+      if verbose:
+        logging.info(
+            "{} of {} masks added,"
+            "attr_sum: {}, area: {:.3g}/{:.3g}, {} remaining masks".format(
+                added_masks_cnt, n_masks, current_attr_sum, current_area_perc,
+                area_perc_th, len(remaining_masks)))
+      added_masks_cnt += 1
+
+    uncomputed_mask = output_attr == -np.inf
+    # Assign the uncomputed areas a value such that sum is same as ig
+    output_attr[uncomputed_mask] = gain_fun(uncomputed_mask, attr)
+    # Set uncomputed region's rank to max rank + 1
+    masks_trace = list(zip(*sorted(masks_trace, key=lambda x: -x[1])))[0]
+    # masks_trace = list(zip(*masks_trace))[0]
+    if integer_segments:
+      attr_ranks = np.zeros(shape=attr.shape, dtype=np.int)
+      for i, mask in enumerate(masks_trace):
+        attr_ranks[mask] = i + 1
+      assert np.all(uncomputed_mask == (attr_ranks == 0))
+      attr_ranks[uncomputed_mask] = np.max(attr_ranks) + 1
+      return output_attr, attr_ranks
+    else:
+      return output_attr, masks_trace
```

Comment:
Should 'masks_trace' contain the segment that contains union of all segments that weren't included? That would make it consistent with the integer mask.

Reply:
done. Appended uncomputed_mask to the end of the list.